File: rb101/rb101-rb109-small_problems/easy9/grocery_list.rb
def buy_fruit(fruits)
  fruits.each_with_object([]) do |(fruit, quantity), list|
    quantity.times { list << fruit }
  end
end

p buy_fruit([["apples", 3], ["orange", 1], ["bananas", 2]]) ==
  ["apples", "apples", "apples", "orange", "bananas", "bananas"]

# ALTERNATIVE SOLUTION:

# 1)
# def buy_fruit(list)
#   expanded_list = []

#   list.each do |item|
#     fruit, quantity = item[0], item[1]
#     quantity.times { expanded_list << fruit }
#   end

#   expanded_list
# end

# 2)
# def buy_fruit(list)
#   list.map { |fruit, quantity| [fruit] * quantity }.flatten
# end

# 3)
# def buy_fruit(fruits)
#   list = []
#   fruits.each do |fruit|
#     fruit[1].times { list << fruit[0] }
#   end
#   list
# end

File: js210/js210-small_problems/list_processing/groceryList.js
function buyFruit(fruitsList) {
  const fruits = [];
  fruitsList.forEach(item => {
    for (let count = 0; count < item[1]; count += 1) {
      fruits.push(item[0]);
    }
  });

  return fruits;
}

console.log(buyFruit([['apple', 3], ['orange', 1], ['banana', 2]]));
// returns (as shown by console output): ["apple", "apple", "apple", "orange", "banana", "banana"]


// ALTERNATIVE SOLUTIONS:

// 1) - LS Solution:
// function buyFruit(fruitsList) {
//   return fruitsList.map(fruit => repeat(fruit))
//                    .reduce((groceryList, fruit) => groceryList.concat(fruit));
// }

// function repeat(fruit) {
//   const result = [];

//   for (let i = 0; i < fruit[1]; i += 1) {
//     result.push(fruit[0]);
//   }

//   return result;
// }

// 2) - LS Solution with flat() variation:
// function buyFruit(fruitsList) {
//   return fruitsList.map(fruit => repeat(fruit)).flat();
// }

// function repeat(fruit) {
//   const result = [];

//   for (let i = 0; i < fruit[1]; i += 1) {
//     result.push(fruit[0]);
//   }

//   return result;
// }

// 3) - Solution inspired by classmate's implementation (for exploration purposes):
// function buyFruit(fruitsList) {
//   return fruitsList.flatMap(([fruit, amount]) => new Array(amount).fill(fruit));
// }
